### DISCUSSION TITLE
Handling an interrupted download

### DISCUSSION THREAD

**Discussion by neal3000** [onevcat/Kingfisher#2223] [General]
I've run into a situation a few times where I get a partial download of an image if the connection gets interrupted - usually by a user going into an elevator. The partial image (usually the top half is ok and the bottom is just black ) stays in the cache and is used in the future.

Is there anyway  to detect that situation as the image is loaded initially ?, or later on to check if there is a 'bad' image?
or something to adjust to avoid it ( increase retries, longer timeout, etc.. )

Looking for suggestions.

Given the scenario this is not an easily reproducible problem. 

Thanks

**Comment by onevcat** [comment]
Ummm. That is quite unexpected, since Kingfisher won't cache the file unless the `URLSession` gives a non-error response in its `urlSession(_:task:didCompleteWithError:)` delegate method. So in your case, it would seem that `URLSession` did not give an error when the network connection is lost, which is a bit surprising.

If that is true, I guess there is no good way to detect it, since from the beginning it is the `URLSession` gives a wrong assumption. I will try to see if I can reproduce this issue in my environment. I believe it is the first step we can say any further words on it.

**Reply by neal3000** [reply]
I forgot ( as in I wrote that piece of code so log ago I forgot ) That I am using my own ImageDataProvider since I am retrieving from Amazon S3.
My code is
public struct S3ImageProvider: ImageDataProvider {
 public func data(handler: @escaping (Result<Data, Error>) -> Void) {
 Task {
            do {
                let s3 = S3Helpers(client: s3Client)  // AWS code ,  a = client.getObject(input: input);   return a.body.readData()
                let data = try  await s3.readFile(bucket: bucket, key: key) 
                handler(.success(data))
            } catch {
                handler(.failure(error))
                //print(error)
                os.Logger.kfext.warning("S3ImageProvider data() caught execption bucket: \(bucket), key: \(key) error: \(error)")
            }

I have logging, but since this has never happened in a development environment. I don't know if there was an exception. 

I classify this as annoying as opposed to serious or show stopper. It's happened to me once or twice in 3 months. And to a beta tester once last week.

**Reply by onevcat** [reply]
Then it should be related on how the `readFile` is implemented. Maybe it gives a normal flow instead of throwing an error even when a timeout happens (with partial data...). I am not using Amazon S3 and its SDK/API at all, so I am not quite sure what can happen in this case. If there is a way to check and identify this case happening, you can then just call `handler` with a failure, then Kingfisher won't try to use or cache the corrupted image.

**Reply by neal3000** [reply]
Thanks - I looked at the AWS code, couldn't find anything obviously wrong. There are code paths that return an empty Data(), but not a partial filled one. 
I'll do some more digging.
